### PR TITLE
Feature/40 add layerlist widget

### PR DESCRIPTION
**Pull request by seahro**
Adds: layerlist widget to widgets page, layerlist sample page to navigation.

 reference gissue #40

### CODE REVIEW

**Comment by TimPurdum** [comment]
Added Compass Button to Widgets - used to reset orientation.

File: samples/dymaptic.GeoBlazor.Core.Sample.Shared/Pages/LayerLists.razor [hunk 1]
```diff
@@ -0,0 +1,15 @@
+﻿@page "/layerLists"
+<PageTitle>LayerLists</PageTitle>
+<h1>LayerLists</h1>
+
+<p class="instructions">
```
Comment (TimPurdum) [inline samples/dymaptic.GeoBlazor.Core.Sample.Shared/Pages/LayerLists.razor:5]: Add an `<a>` tag that links to the JS API page for LayerLists. See other sample pages for example.

File: samples/dymaptic.GeoBlazor.Core.Sample.Shared/Pages/Widgets.razor [hunk 2]
```diff
@@ -36,6 +36,9 @@
         <div class="form-group">
             <label>Compass: <input type="checkbox" @onchange="@(() => ToggleWidget(nameof(CompassWidget)))"></label>
         </div>
+        <div class="form-group">
```
Comment (TimPurdum) [inline samples/dymaptic.GeoBlazor.Core.Sample.Shared/Pages/Widgets.razor:39]: Per discussion, this page is getting crowded, let's revert these changes, keep `LayerList` just on the new page.

File: samples/dymaptic.GeoBlazor.Core.Sample.Shared/Shared/NavMenu.razor [hunk 3]
```diff
@@ -55,6 +55,11 @@
                 <span class="oi oi-arrow-right" aria-hidden="true"></span> Vector Layer
             </NavLink>
         </div>
+        <div class="nav-item px-3">
+            <NavLink class="nav-link" href="layerLists">
+                <span class="oi oi-layers" aria-hidden="true"></span> Layer Lists
```
Comment (TimPurdum) [inline samples/dymaptic.GeoBlazor.Core.Sample.Shared/Shared/NavMenu.razor:60]: I had already used `oi-layers` for feature-layers above. See if you can find an appropriate different icon in the openiconic set. It can be relatively random, just should be unique. At some point we may want to replace these with custom icons.

File: src/dymaptic.GeoBlazor.Core.sln [hunk 4]
```diff
@@ -13,7 +13,7 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "dymaptic.GeoBlazor.Core.Sam
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "dymaptic.GeoBlazor.Core.Sample.Maui", "..\samples\dymaptic.GeoBlazor.Core.Sample.Maui\dymaptic.GeoBlazor.Core.Sample.Maui.csproj", "{08193B44-3A32-4836-96A0-44DEEA1E7556}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "dymaptic.GeoBlazor.Core.Test", "..\tests\dymaptic.GeoBlazor.Core.Test\dymaptic.GeoBlazor.Core.Test.csproj", "{48E045E3-F3FF-4D50-8987-83C1B8663E39}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "dymaptic.GeoBlazor.Core.Test", "..\tests\dymaptic.GeoBlazor.Core.Test\dymaptic.GeoBlazor.Core.Test.csproj", "{48E045E3-F3FF-4D50-8987-83C1B8663E39}"
```
Comment (TimPurdum) [inline src/dymaptic.GeoBlazor.Core.sln:16]: I'm not sure why this changed in the solution file? It's possibly something VS is updating behind the scenes. I'd try reverting it, re-open VS, rebuild the solution, and see if it does it again. If so, leave the change.

File: src/dymaptic.GeoBlazor.Core/Components/Widgets/LayerList/ActionSection.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace dymaptic.GeoBlazor.Core.Components.Widgets.LayerList;
+
+public class ActionSection
+{
+    /// <summary>
```
Comment (TimPurdum) [inline src/dymaptic.GeoBlazor.Core/Components/Widgets/LayerList/ActionSection.cs:11]: This summary xml doc should be moved to the line above `public class ActionSection` so that it applies to the whole class, not the `Title` property. Also, as a formatting suggestion, I like to indent my xml comments like real xml/html, e.g.

```csharp
/// <summary>
///     this is the content
/// </summary>
```

File: src/dymaptic.GeoBlazor.Core/Components/Widgets/LayerListWidget.cs [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    public override string WidgetType => "layerList";
+
+    [Parameter]
+    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
+    public string? IconClass { get; set; }
+    [Parameter]
+    public string? Label { get; set; }
+    public Func<ListItem, Task<ListItem>>? OnListItemSelectedHandler { get; set; } 
+    public DotNetObjectReference<LayerListWidget> LayerListWidgetObjectReference => DotNetObjectReference.Create(this);
+    [JSInvokable]
```
Comment (TimPurdum) [inline src/dymaptic.GeoBlazor.Core/Components/Widgets/LayerListWidget.cs:26]: Let's add some blank line breathing room, especially when moving from properties to methods.

File: src/dymaptic.GeoBlazor.Core/Scripts/arcGisJsInterop.ts [hunk 7]
```diff
@@ -1075,4 +1110,32 @@ function waitForRender(viewId: string, dotNetRef: any): void {
...
+    if (item === undefined || item === null) return null;
+    let children: Array<DotNetListItem> = [];
+    item.children.forEach(c => {
+        let child = buildDotNetListItem(c);
+        if (child !== null) {
+            children.push(child);
+        }
+    });
+
+    /// <summary>
```
Comment (TimPurdum) [inline src/dymaptic.GeoBlazor.Core/Scripts/arcGisJsInterop.ts:1127]: I don't think this comment is needed here. Our library consumers will never call this code directly. It is only called by _our_ code above.

File: src/dymaptic.GeoBlazor.Core/Scripts/arcGisJsInterop.ts [hunk 8]
```diff
@@ -1075,4 +1110,32 @@ function waitForRender(viewId: string, dotNetRef: any): void {
...
+    /// and toggle the selected layer's visiblity. More capabilities will be available after full
+    /// implementation of ActionSection.
+    /// </summary>
+
+    return {
+        title: 'title',
+        layer: item.layer,
+        visible: item.visible,
+        children: children,
+        actionSections: item.actionsSections
```
Comment (TimPurdum) [inline src/dymaptic.GeoBlazor.Core/Scripts/arcGisJsInterop.ts:1139]: Likewise, I still believe that `item.actionsSections` here will always be null or empty, bc there's been no chance to populate it. It might be worth putting a breakpoint there to see if I'm wrong.

File: src/dymaptic.GeoBlazor.Core/Scripts/definitions.d.ts [hunk 9]
```diff
@@ -1,4 +1,7 @@
-﻿declare module "ArcGisDefinitions" {
+﻿import Layer from "@arcgis/core/layers/Layer";
+import ListItem from "@arcgis/core/widgets/LayerList/ListItem";
```
Comment (TimPurdum) [inline src/dymaptic.GeoBlazor.Core/Scripts/definitions.d.ts:2]: this import is no longer used

**Review by TimPurdum — CHANGES_REQUESTED** [review]
Nice work on this! A few things to clean up, then I'll re-review tomorrow.

File: samples/dymaptic.GeoBlazor.Core.Sample.Shared/Pages/LayerLists.razor (hunk 1 above)
Comment (seahro) [reply samples/dymaptic.GeoBlazor.Core.Sample.Shared/Pages/LayerLists.razor:5]: corrected

File: samples/dymaptic.GeoBlazor.Core.Sample.Shared/Pages/Widgets.razor (hunk 2 above)
Comment (seahro) [reply samples/dymaptic.GeoBlazor.Core.Sample.Shared/Pages/Widgets.razor:39]: corrected

File: samples/dymaptic.GeoBlazor.Core.Sample.Shared/Shared/NavMenu.razor (hunk 3 above)
Comment (seahro) [reply samples/dymaptic.GeoBlazor.Core.Sample.Shared/Shared/NavMenu.razor:60]: list icon inputted

File: src/dymaptic.GeoBlazor.Core/Components/Widgets/LayerList/ActionSection.cs (hunk 5 above)
Comment (seahro) [reply src/dymaptic.GeoBlazor.Core/Components/Widgets/LayerList/ActionSection.cs:11]: adjusted and reformatted

File: src/dymaptic.GeoBlazor.Core/Components/Widgets/LayerList/ListItem.cs [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+
+public class ListItem 
+{
+    [Parameter]
+    public string? Title { get; set; }
+    [Parameter]
+    public Layer? Layer { get; set; }
+    [Parameter]
+    public bool? Visible { get; set; }
+    public List<ListItem>? Children { get; set; }
```
Comment (TimPurdum) [inline src/dymaptic.GeoBlazor.Core/Components/Widgets/LayerList/ListItem.cs:23]: I'm a big fan of using whitespace (blank lines) to make things easier to visualize/read. I'm not 100% consistent about it, but in general, I find it better to add space above attributes and xml comments.

**Review by TimPurdum — APPROVED** [review]
Well done, @seahro! I made a few spacing/comment tweaks, and will merge for you (so I can pull into my PR this morning).
